File: Binary-Search/0154-find-minimum-in-rotated-sorted-array-ii.py
# Time  : O(logn)
# Space : O(1)

class Solution:
  def findMin(self, A: List[int]) -> int:
    lo, hi = 0, len(A) - 1

    while lo < hi:
      mid = lo + (hi - lo) // 2
      if A[mid] > A[hi]:
        lo = mid + 1
      elif A[mid] < A[hi]:
        hi = mid
      else:  # A[mid] == A[hi]
        if hi > 0 and A[hi - 1] <= A[hi]:
          hi -= 1
        else:
          break

    return A[hi]

File: Binary-Search/0154-find-minimum-in-rotated-sorted-array-ii.cpp
// Time  : O(logn)
// Space : O(1)
class Solution {
public:
  int findMin(vector<int>& A) {
    int lo = 0;
    int hi = A.size()-1;
    while (lo < hi) {
      int mid = (lo + hi)>>1;
      if (A[mid] > A[hi]) lo = mid+1;
      else if (A[mid] < A[hi]) hi = mid;
      else {
        if (A[hi-1] <= A[hi]) hi--;
        else {
          lo = hi;
          break;
        }
      }
    }
    return A[lo];
  }
};